### PR TITLE
Raise a warning when applying a pulse gradient transform to a QNode directly

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
Applying a `gradient_transform` to QNodes with non-scalar parameters directly does not work correctly (see #4225).

**Description of the Change:**
This PR introduces a comment in the docstrings of `pulse_generator` and `stoch_pulse_grad` and makes them raise a warning when applied to a QNode directly. The warning is raise always, even for scalar parameters, which is in order to avoid too much code modifications.

**Benefits:**
Clarity about failing gradient computations and warnings for wrong or misshaped results.

**Possible Drawbacks:**
Warning raised in cases that are actually fine. However, users could/should generically move towards using JAX entrypoints for pulses instead.

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/gradients/pulse_gradient.py [hunk 1]
```diff
@@ -307,10 +320,11 @@ def _stoch_pulse_grad(
         rules when used with simple pulses (see details and examples below), potentially leading
         to imprecise results and/or unnecessarily large computational efforts.
 
-    .. note::
+    .. warning::
 
-        Currently this function only supports pulses for which each *parametrized* term is a
-        simple Pauli word. More general Hamiltonian terms are not supported yet.
```
Comment (dwierichs) [inline pennylane/gradients/pulse_gradient.py:313]: This note was outdated and should have been removed in #4132 .

File: tests/gradients/core/test_pulse_gradient.py [hunk 2]
```diff
@@ -803,8 +803,41 @@ def test_shots_attribute(self, shots):
...
+        with qml.Tracker(dev) as tracker:
+            _match = "stochastic pulse parameter-shift .* scalar pulse parameters."
+            with pytest.warns(UserWarning, match=_match):
+                grad_fn = stoch_pulse_grad(circuit, num_split_times=2)
+            grad = grad_fn(params)
+
+        p = params * T
+        exp_grad = -2 * jnp.sin(2 * p) * T
+        assert jnp.allclose(grad, exp_grad)
+        assert tracker.totals["executions"] == 4  # two shifted tapes, two splitting times
```
Comment (dwierichs) [inline tests/gradients/core/test_pulse_gradient.py:835]: This is a copy of the test for `pulse_generator`, just to also have at least one test on the `QNode` level for `stoch_pulse_grad`. As it has scalar parameters, it actually works properly. However, as we always raise the warning, we can also check here for the warning being raised.

File: pennylane/gradients/pulse_generator_gradient.py [hunk 3]
```diff
@@ -702,3 +709,15 @@ def expand_invalid_trainable_pulse_generator(x, *args, **kwargs):
...
+@pulse_generator.custom_qnode_wrapper
+def pulse_generator_qnode_wrapper(self, qnode, targs, tkwargs):
+    """A custom QNode wrapper for the gradient transform :func:`~.pulse_generator`.
+    It is equivalent to the default QNode wrapper of any :func:`~.gradient_transform`,
+    but in addition warns that applying ``pulse_generator`` to a ``QNode`` directly
+    is only supported for scalar pulse parameters.
+    """
+    transform_name = "pulse generator parameter-shift"
+    warn_pulse_diff_on_qnode(transform_name)
+    return self.default_qnode_wrapper(qnode, targs, tkwargs)
```
Comment (trbromley) [inline pennylane/gradients/pulse_generator_gradient.py:723]: Will this warning come up when the transform is used as part of the JAX entrypoints like `jax.grad`?

Comment (dwierichs) [reply pennylane/gradients/pulse_generator_gradient.py:723]: No, because these entrypoints use the tape transform. (That's also why they actually support non-scalar parameters already :))

**Review by Qottmann — APPROVED** [review]
Thanks for the upgrade @dwierichs 👌 
I agree with the notion to recommend usage of jax entrypoints and appreciate the explicit mentioning in the docs and warning messages 👍 
Took the liberty to fix one test, hope that is okay 😅 Approving assuming passing tests now :)

**Comment by dwierichs** [comment]
> Took the liberty to fix one test, hope that is okay sweat_smile Approving assuming passing tests now :)

oh great, thanks for catching my slopiness there! :)

File: doc/releases/changelog-dev.md [hunk 4]
```diff
@@ -43,6 +43,11 @@
 
 <h3>Improvements 🛠</h3>
 
+* The pulse differentiation methods, `pulse_generator` and `stoch_pulse_grad` now raise a warning when they
+  are applied to a `QNode` directly, which is only supported for scalar pulse parameters. For non-scalar
+  pulse parameters, use differentiation via a JAX entry point (`jax.grad`, `jax.jacobian`, ...) instead
+  [(4241)](https://github.com/PennyLaneAI/pennylane/pull/4241)
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:49]:
```suggestion
* The pulse differentiation methods, `pulse_generator` and `stoch_pulse_grad` now raise an error when they
  are applied to a `QNode` directly. Instead, use differentiation via a JAX entry point (`jax.grad`, `jax.jacobian`, ...).
  [(4241)](https://github.com/PennyLaneAI/pennylane/pull/4241)
```

File: pennylane/gradients/pulse_generator_gradient.py [hunk 5]
```diff
@@ -455,7 +455,14 @@ def _pulse_generator(tape, argnum=None, shots=None, atol=1e-7):
...
         commonly encountered with PennyLane.
-        In addition, this transform is only JIT-compatible with pulses that only have scalar parameters.
+        In addition, this transform is only JIT-compatible with pulses that only have scalar
+        parameters.
+
+    .. warning::
+
+        This transform may only be applied directly to QNodes that only have scalar parameters.
+        For non-scalar parameters, use JAX entrypoints (``jax.grad``, ``jax.jacobian``, ...)
+        instead or apply the transform on the tape level. Also see the examples below.
```
Comment (dwierichs) [inline pennylane/gradients/pulse_generator_gradient.py:465]:
```suggestion
        This transform may not be applied directly to QNodes. Use JAX entrypoints 
        (``jax.grad``, ``jax.jacobian``, ...) instead or apply the transform on the tape
        level. Also see the examples below.
```

File: pennylane/gradients/pulse_gradient.py [hunk 6]
```diff
@@ -676,3 +691,15 @@ def expand_invalid_trainable_stoch_pulse_grad(x, *args, **kwargs):
...
     _stoch_pulse_grad, expand_fn=expand_invalid_trainable_stoch_pulse_grad
 )
+
+
+@stoch_pulse_grad.custom_qnode_wrapper
+def stoch_pulse_grad_qnode_wrapper(self, qnode, targs, tkwargs):
+    """A custom QNode wrapper for the gradient transform :func:`~.stoch_pulse_grad`.
+    It is equivalent to the default QNode wrapper of any :func:`~.gradient_transform`,
+    but in addition warns that applying ``stoch_pulse_grad`` to a ``QNode`` directly
+    is only supported for scalar pulse parameters.
```
Comment (dwierichs) [inline pennylane/gradients/pulse_gradient.py:701]:
```suggestion
    It raises an error, so that applying ``pulse_generator`` to a ``QNode`` directly
    is not supported.
```

File: pennylane/gradients/pulse_gradient.py [hunk 7]
```diff
@@ -56,6 +56,20 @@ def _assert_has_jax(transform_name):
         )
 
 
+def raise_pulse_diff_on_qnode(transform_name):
+    """Emit a warning that the gradient transform with the provided name only supports
+    direct application to QNodes that have scalar pulse parameters.
```
Comment (dwierichs) [inline pennylane/gradients/pulse_gradient.py:61]:
```suggestion
    """Raises an error as the gradient transform with the provided name does
    not support direct application to QNodes.
```

File: pennylane/gradients/pulse_gradient.py [hunk 8]
```diff
@@ -56,6 +56,20 @@ def _assert_has_jax(transform_name):
...
+    """
+    msg = (
+        f"Applying the {transform_name} gradient transform to a QNode directly is currently "
+        "only supported for scalar pulse parameters. Non-scalar parameters may lead to wrong "
+        "results or raise exceptions. Please use differentiation via a JAX entry point "
+        "(jax.grad, jax.jacobian, ...) instead for non-scalar parameters.",
```
Comment (dwierichs) [inline pennylane/gradients/pulse_gradient.py:67]:
```suggestion
        f"Applying the {transform_name} gradient transform to a QNode directly is currently "
        "not supported. Please use differentiation via a JAX entry point "
        "(jax.grad, jax.jacobian, ...) instead.",
```

File: pennylane/gradients/pulse_gradient.py [hunk 9]
```diff
@@ -307,10 +321,11 @@ def _stoch_pulse_grad(
...
         to imprecise results and/or unnecessarily large computational efforts.
 
-    .. note::
+    .. warning::
 
-        Currently this function only supports pulses for which each *parametrized* term is a
-        simple Pauli word. More general Hamiltonian terms are not supported yet.
+        This transform may only be applied directly to QNodes that only have scalar parameters.
+        For non-scalar parameters, use JAX entrypoints (``jax.grad``, ``jax.jacobian``, ...)
+        instead or apply the transform on the tape level. Also see the examples below.
```
Comment (dwierichs) [inline pennylane/gradients/pulse_gradient.py:328]:
```suggestion
        This transform may not be applied directly to QNodes. Use JAX entrypoints
         (``jax.grad``, ``jax.jacobian``, ...)
        instead or apply the transform on the tape level. Also see the examples below.
```

**Comment by dwierichs** [comment]
Thanks @trbromley, I changed the texts. :)

**Comment by trbromley** [comment]
This can be merged during next week as it is more of a bug fix
